File: Silver/2017-03/pairup_open2017.java
// Paired Up - USACO Silver US Open 2017 (http://www.usaco.org/index.php?page=viewproblem2&cpid=738)
// This problem was completed as classwork for the USACO Silver 2 Class on 7/10/21.

import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class L19_pairup {
    static class CowGroup implements Comparable<CowGroup> {
        int unpairedCows;
        int milkTime;

        CowGroup(int unpairedCows, int milkTime) {
            this.unpairedCows = unpairedCows;
            this.milkTime = milkTime;
        }

        @Override
        public String toString() {
            return "(" + unpairedCows + ", " + milkTime + ")";
        }

        @Override
        public int compareTo(CowGroup o) {
            return milkTime - o.milkTime;
        }
    }

    public static void main(String[] args) throws IOException {
        // input
        String problemName = "pairup";
        BufferedReader br = new BufferedReader(new FileReader(problemName + ".in"));
        StringTokenizer st;

        int numLines = Integer.parseInt(br.readLine());
        CowGroup[] cows = new CowGroup[numLines];
        for (int i=0; i<numLines; i++) {
            st = new StringTokenizer(br.readLine());
            cows[i] = new CowGroup(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }

        // algorithm
        Arrays.sort(cows);

        int left = 0;
        int right = numLines-1;
        int maxMilkingTime = 0;

        // greedy algorithm
        while (left <= right) {
            maxMilkingTime = Math.max(maxMilkingTime, cows[left].milkTime + cows[right].milkTime);

            if (cows[right].unpairedCows > cows[left].unpairedCows) {
                // pair all left cows with as many right cows as possible
                cows[right].unpairedCows -= cows[left].unpairedCows;
                left++;
            } else if (cows[left].unpairedCows > cows[right].unpairedCows) {
                // pair all right cows with as many left cows as possible
                cows[left].unpairedCows -= cows[right].unpairedCows;
                right--;
            } else {
                // pair all left cows with all right cows
                left++;
                right--;
            }
        }

        // output
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter(problemName + ".out")));

        pw.println(maxMilkingTime);

        br.close();
        pw.close();
    }
}

File: Silver/2017-03/pairup_open2017.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int, int> pii;

int n, m = 0;
pii cows[100001];

int main() {
    ifstream fin("pairup.in");
    ofstream fout("pairup.out");

    fin >> n;
    for (int i=1; i<=n; i++) {
        int x, y; fin >> x >> y;
        cows[i] = pii(y, x);
        m += x;
    }

    sort(cows+1, cows+n+1);

    int result = 0;
    int left = 1, right = n;

    while (left <= right) {
        int& leftCows = cows[left].second;
        int& rightCows = cows[right].second;

        result = max(result, cows[left].first + cows[right].first);

        if (leftCows == rightCows) {
            left++;
            right--;
        } else if (leftCows < rightCows) {
            left++;
            rightCows -= leftCows;
        } else {
            right--;
            leftCows -= rightCows;
        }
    }

    fout << result << endl;
}